_drafts/practical_zephyr_03_dt_semantics.md

```diff
@@ -0,0 +1,1970 @@
+
+
+### Labels and paths
+
+Before we move on to properties using `phandle` types, let's have a look at labels and paths. In the previous article, we've seen that the [Devicetree specification](https://www.devicetree.org/specifications/) allows _labels_ not only to refer to _nodes_ but also to other positions in the Devicetree, e.g., in front of property _values_. In our example, we have three labels:
+
+- One label `label_with_props` for our node `/node_with_props`,
+- Two labels `second_value` and `string_value` for some property values.
+
+`dts/playground/props-basics.overlay`
+```dts
+/ {
+  label_with_props: node_with_props {
+    compatible = "custom-props-basics";
+    /* ... */
+    array = <1 second_value: 2 3>;
+    /* ... */
+    string = string_value: "foo bar baz";
+    string-array = "foo", "bar", "baz";
+    /* ... */
+  };
+};
+```
+
+Zephyr's DTS generator accepts this input and - as we've seen - won't complain about the labels `second_value` and `string_value` referring to the property values. However, Zephyr simply ignores any label that doesn't refer to a _node_ and therefore doesn't generate anything usable for the property labels. Only the _node_ label `label_with_props` has a macro:
+
+```bash
+$ grep second_value ../build/zephyr/include/generated/devicetree_generated.h
+$ grep string_value ../build/zephyr/include/generated/devicetree_generated.h
+$ grep label_with_props ../build/zephyr/include/generated/devicetree_generated.h
+#define DT_N_NODELABEL_label_with_props DT_N_S_node_with_props
+```
+
+> **Note:** In case you want some practice, look at the [Devicetree specification](https://www.devicetree.org/specifications/) and try to use references to value labels in your Devicetree. I promise it won't work either.
+
+The following macros are generated to represent a node and its labels:
+
+`build/zephyr/include/generated/devicetree_generated.h`
+```c
+/*
+ * Devicetree node: /node_with_props
+ *
+ * Node identifier: DT_N_S_node_with_props
+ * --snip--
+ */
+/* Node's full path: */
+#define DT_N_S_node_with_props_PATH "/node_with_props"
+/* Node's name with unit-address: */
+#define DT_N_S_node_with_props_FULL_NAME "node_with_props"
+// --snip--
+/* Existence and alternate IDs: */
+#define DT_N_S_node_with_props_EXISTS 1
+#define DT_N_INST_0_custom_props_basics DT_N_S_node_with_props
+#define DT_N_NODELABEL_label_with_props DT_N_S_node_with_props
+```
+
+The _identifier_ that is used for our `/node_with_props` is `DT_N_S_node_with_props`, which is exactly what we'd expect knowing the node's path and from what we've learned about [understanding Devicetree macro names](#understanding-devicetree-macro-names). This _identifier_ is itself just a _token_: No macro with the name `DT_N_S_node_with_props` exists!
```

Comment:
I noticed that later on the name "node identifier" is used and I think it would be good to switch identifier -> node identifier as necessary

Reply:
Yep, makes sense. I'll change it.